tsl/src/compression/compression.c

```diff
@@ -323,14 +346,180 @@ compress_chunk(Oid in_table, Oid out_table, const ColumnCompressionInfo **column
+	 *
+	 * Matching Criteria for Each IndexAtt[i] and ColumnCompressionInfo Keys[i]
+	 * ========================================================================
+	 * a) Index attnum must match with ColumnCompressionInfo Key {keys[i]}.
+	 * b) Index attOption(ASC/DESC and NULL_FIRST) can be mapped with ColumnCompressionInfo
+	 * orderby_asc and null_first.
+	 *
+	 * BTREE Indexes Ordering
+	 * =====================
+	 * a) ASC[Null_Last] ==> [1]->[2]->NULL
+	 * b) [Null_First]ASC ==> NULL->[1]->[2]
+	 * c) DSC[Null_Last]  ==> [2]->[1]->NULL
+	 * d) [Null_First]DSC ==> NULL->[2]->[1]
+	 */
+	if (ts_guc_enable_compression_indexscan)
+	{
+		foreach (lc, in_rel_index_oids)
+		{
+			Oid index_oid = lfirst_oid(lc);
+			Relation index_rel = index_open(index_oid, AccessShareLock);
+			IndexInfo *index_info = BuildIndexInfo(index_rel);
+			int previous_direction = NoMovementScanDirection;
+			int current_direction = NoMovementScanDirection;
 
-	Tuplesortstate *sorted_rel = compress_chunk_sort_relation(in_rel, n_keys, keys);
+			if (n_keys <= index_info->ii_NumIndexKeyAttrs && index_info->ii_Am == BTREE_AM_OID)
+			{
+				for (i = 0; i < n_keys; i++)
+				{
+					int16 att_num = get_attnum(in_table, NameStr(keys[i]->attname));
+
+					int16 option = index_rel->rd_indoption[i];
+					bool index_orderby_asc = ((option & INDOPTION_DESC) == 0);
+					bool index_null_first = ((option & INDOPTION_NULLS_FIRST) != 0);
+					bool is_orderby_asc = keys[i]->orderby_asc;
+					bool is_null_first = keys[i]->orderby_nullsfirst;
+
+					if (att_num > 0 && index_info->ii_IndexAttrNumbers[i] == att_num)
+					{
+						in_table_tp = SearchSysCacheAttNum(in_table, att_num);
+						if (!HeapTupleIsValid(in_table_tp))
+							elog(ERROR,
+								 "table \"%s\" does not have column \"%s\"",
+								 get_rel_name(in_table),
+								 NameStr(keys[i]->attname));
+
+						index_tp = SearchSysCacheAttNum(index_oid, i + 1);
+						if (!HeapTupleIsValid(index_tp))
+							elog(ERROR,
+								 "index \"%s\" does not have column \"%s\"",
+								 get_rel_name(index_oid),
+								 NameStr(keys[i]->attname));
+
+						in_table_attr_tp = (Form_pg_attribute) GETSTRUCT(in_table_tp);
+						index_attr_tp = (Form_pg_attribute) GETSTRUCT(index_tp);
+
+						if (COMPRESSIONCOL_IS_SEGMENT_BY(keys[i]))
+						{
+							/* Segment_by column is always ASC in compressed chunk/s
+							 * regardless of order_by column's ordering. This is a
+							 * major distinction between ordering in indexes and
+							 * segmented-ordering in compressed chunks. If a key is
+							 * segment_by in compression and ASC in index then the
+							 * direction will be forward otherwise backward to keep
+							 * the segment_by always ASC.
+							 */
+							if (option)
+							{
+								if (!index_orderby_asc &&
+									in_table_attr_tp->attcollation == index_attr_tp->attcollation)
+								{
+									current_direction = BackwardScanDirection;
+								}
+								else if (in_table_attr_tp->attcollation ==
```

Comment:
Codecov complains that there's no test coverage for this and the next condition, let's add it.

Reply:
Now coverage looks good